src/lib/utils/thorchain/memo/addFinalAssetLimitToMemo.ts

```diff
@@ -0,0 +1,70 @@
+import { assertAndProcessMemo } from './assertAndProcessMemo'
+import { MEMO_PART_DELIMITER } from './constants'
+
+type AddLimitToMemoArgs = {
+  /**
+   * A THORChain memo, with or without a limit/MinOut.
+   */
+  memo: string
+  /**
+   * The final asset amount out to be added to the memo. THOR Aggregator will use the last two numbers as exponents.
+   */
+  finalAssetAmountOut: string
+}
+
+/**
+ * A simple util that adds a final asset limit to a THORChain memo, if the memo doesn't already have one.
+ */
+export const addFinalAssetLimitToMemo = ({ memo, finalAssetAmountOut }: AddLimitToMemoArgs) => {
+  const memoParts = memo.split(MEMO_PART_DELIMITER)
+  const [action] = memoParts
+
+  if (!action) throw new Error(`action is required in memo: ${memo}`)
+
+  let updatedMemo = memo
+
+  switch (action.toLowerCase()) {
+    case 'swap':
+    case '=':
+    case 's': {
+      // SWAP:ASSET:DESTADDR:LIM/INTERVAL/QUANTITY:AFFILIATE:FEE:AGGREGATOR:FINALASSETCONTRACT:FINALASSETAMOUNTOUT
+      const [
+        ,
+        asset,
+        destAddr,
+        limit,
+        affiliate,
+        fee,
+        aggregator,
+        finalAssetContract,
+        _finalAssetAmountOut,
+      ] = memoParts
+      if (!_finalAssetAmountOut) {
+        updatedMemo = `${action}:${asset}:${destAddr}:${limit}:${affiliate || ''}:${
+          fee || ''
+        }:${aggregator}:${finalAssetContract}:${finalAssetAmountOut}`
+      }
+      break
+    }
+    // No final asset amount out for loans as of now
+    case '$+':
+    case 'loan+':
+      break
+    // No final asset amount out for loans as of now
+    case '$-':
+    case 'loan-':
+      break
+    // No final asset amount out for Add Liquidity / Withdraw Liquidity as of now
+    case 'add':
+    case '+':
+    case 'a':
+    case 'withdraw':
+    case '-':
+    case 'wd':
+      break
```

Comment:
Question: Should we throw rather than no-op for all these as this shouldn't be called on memos other than SwapOut (same q. for all other utils)

Reply:
Why not yep, added some throws in [2a6fc6a](https://github.com/shapeshift/web/pull/7146/commits/2a6fc6a57521f17fb2fa9ee67d3d46f222c6e53c)